pyquil/api/qpu.py

```diff
@@ -120,38 +123,45 @@ def _run_payload(self, quil_program, classical_addresses, trials):
 
         payload = {"type": TYPE_MULTISHOT,
                    "addresses": list(classical_addresses),
-                   "trials": trials,
-                   "quil-instructions": quil_program.out()}
+                   "trials": trials}
+
+        if needs_compilation:
+            payload["uncompiled-quil"] = quil_program.out()
+            if isa:
+                payload["isa"] = isa.to_dict()
+        else:
+            payload["compiled-quil"] = quil_program.out()
 
         return payload
 
-    def run_and_measure(self, quil_program, qubits, trials=1):
+    def run_and_measure(self, quil_program, qubits, trials=1, needs_compilation=True, isa=None):
         """
         Run a pyQuil program on the QPU multiple times, measuring all the qubits in the QPU
         simultaneously at the end of the program each time. This functionality is in beta.
 
         :param Program quil_program: A Quil program.
         :param list|range qubits: The list of qubits to measure
         :param int trials: Number of shots to collect.
+        :param needs_compilation: If True, preprocesses the job with the compiler.
```

Comment:
So what are the implications? Does this mean that canopy dispatches it to QUILC first and then let's redwood pull it? And is there no compilation done in redwood anymore?

Reply:
Right. The possible execution paths, depending on run method arguments, are:
* A job is sent to quilc over a synchronous connection. (Only option for sync CompilerConnection)
* A job is sent to qvm over a synchronous connection. (Only option for sync QVMConnection)
* A job is sent to the queue over an async connection; it is passed first to quilc and forwarded to the qpu without leaving the cloud, and the qpu software stack performs no further compilation. (Default for qpu)
* A job is sent to the queue over an async connection; it is passed to quilc and immediately returns. (Only option for async CompilerConnection)
* A job is sent to the queue over an async connection; it is passed directly to the qpu.
* A job is sent to the queue over an async connection; it is passed directly to the qvm. (Default for qvm)
* A job is sent to the queue over an async connection; it is passed first to quilc and forwarded to the qvm without leaving the cloud.